GCS direct uploads

fiedosiukr opened the pull request:
## Changes
- Added [Dropzone.js](https://www.dropzone.dev/js/) to handle direct uploads to GCS
- Added conditional template rendering for different storage types
- Created an endpoint for generating signed urls
- Improved some of the JS code

cx1111 committed:
Add django-strages and boto3 packages

grzegorz-wcislo committed:
Update poetry lockfile

grzegorz-wcislo committed:
Add deps to requirements.txt

grzegorz-wcislo committed:
Add deps to requirements.txt

grzegorz-wcislo committed:
Add deps to requirements.txt

grzegorz-wcislo committed:
Bump dependency hash

grzegorz-wcislo committed:
Add creating license files

grzegorz-wcislo committed:
Add creating license files

grzegorz-wcislo committed:
Add creating license files

grzegorz-wcislo committed:
Add creating license files

grzegorz-wcislo committed:
Update file-storage documentation

marcinjosinski committed:
Update dependencies

marcinjosinski committed:
Update dependencies

marcinjosinski committed:
Replace Django Form with Dropzone form when uploading files

marcinjosinski committed:
Add endpoint to generate signed urls

Commit:
Improve signed url generation. Add file size verification, modify templates.

Commit:
Moved dropzone static files to GCP.

Commit:
Modify console view and template.

Commit:
Fix queryset.

grzegorz-wcislo commented on FILE-STORAGE.md:
```diff
@@ -0,0 +1,53 @@
+- User-uploaded media files are stored in the `MEDIA_ROOT` folder. Authorization is done by Django views and then an `X-Accel-Redirect` to nginx if approved.
+- When projects get published, if it is a public project, files get moved to a `published_projects` subdirectory within `STATIC_ROOT`. Otherwise they get moved into a `published_projects` subdirectory within `MEDIA_ROOT`.
+
+## GCS File Storage
```
Could you add info regarding CORS configuration for the media bucket (required for direct uploads)

grzegorz-wcislo commented on physionet-django/console/views.py:
```diff
@@ -411,7 +411,7 @@ def copyedit_submission(request, project_slug, *args, **kwargs):
         subdir = ''
 
-    authors, author_emails, storage_info, edit_logs, copyedit_logs, latest_version = project.info_card()
+    authors, author_emails, storage_info, edit_logs, copyedit_logs, latest_version = project.info_card(force_calculate=True)
```
I'm not sure if we want to use force_calculate when using local file storage here

grzegorz-wcislo commented on physionet-django/project/templates/project/edit_files_panel.html:
```diff
@@ -1,7 +1,20 @@
+{% endblock %}
+
+{% block local_css %}
+<link rel="stylesheet" type="text/css" href="{% static 'dropzone/dropzone.min.css' %}">
```
What's the difference between `dropzone.min.css` and `basic.min.css`? Do we need both?

fiedosiukr replied:
It seems like having `dropzone.min.css` is sufficient.

Commit:
Add error handling, update FILE-STORAGE.md, remove redundant CSS file.

Commit:
Fix file uploads.

Commit:
Fix file uploads.

Commit:
Add test cases.

Commit:
Fix project dependencies.

Commit:
Fix project dependencies.

Commit:
Add missing dependencies to requirements.txt file.

Commit:
Update poetry.lock

Commit:
Update poetry.lock

Commit:
Add allow_field_downloads field.

Commit:
Remove botocore.

Commit:
Fix file_root method.

Commit:
Change underscores to dashes in the url.

Commit:
Move allow_file_downloads to the files section.

Commit:
Remove gcs fake server from workflow, remove redundant dependencies.

Commit:
Remove gcs fake server from workflow, remove redundant dependencies.

Commit:
Move allow_file_downloads filed to the Access section.

Commit:
Fix tests.

Commit:
Remove redundant form.

Commit:
Fix typos.

Commit:
Remove print statements, and s3 related stuff.

Commit:
Create section model.

Commit:
Add new env variable, create custom context processor.

fiedosiukr committed:
Implement section ordering, create section view.

fiedosiukr committed:
Add new menu item, add static pages view.

fiedosiukr committed:
Add dynamic sections rendering.

fiedosiukr committed:
Use PLATFORM_NAME instead of the hard-coded value.

fiedosiukr committed:
Add missing migrations.

fiedosiukr committed:
Add section edit view.

Commit:
Remove unused variable.

Commit:
.env: new option BULK_DOWNLOAD_HOSTNAME.

This option can be used to specify an alternate hostname that will be
used for bulk download commands (currently, 'wget'.)  This hostname
will be used *in place of* the server's hostname, and therefore it
should ultimately be another name for the same server, or a proxy or
mirror that is able to serve the same resources.

However, it's useful to have the option of promoting a different
hostname for downloads than the hostname used for interactive
browsing.

Commit:
get_url_prefix: add bulk_download parameter.

If this optional parameter is set to true, and BULK_DOWNLOAD_HOSTNAME
is defined in the settings, then use that hostname rather than the
current site hostname to construct the URL prefix.

Commit:
published_project: display bulk URL prefix in wget command.

In a published project, the "files" section shows an example wget
command that can be used to download the project.

If BULK_DOWNLOAD_HOSTNAME is defined, we want to use that hostname in
the wget command, rather than the current site hostname.  Do this by
passing a template parameter 'bulk_url_prefix' instead of 'url_prefix'
to avoid confusion.  (Note that 'url_prefix' is not currently used
anywhere else on the published project page, so we can remove the old
template parameter.)

Commit:
submission_info_card.html: display bulk URL prefix in wget command.

In the unpublished project submission info pages, an example wget
command is shown that can be used to download the project.

If BULK_DOWNLOAD_HOSTNAME is defined, we want to use that hostname in
the wget command, rather than the current site hostname.  Do this by
defining a template parameter 'bulk_url_prefix' to avoid confusion.

The following views use submission_info_card.html:
  console.views.submission_info (via console/submission_info.html)
  console.views.edit_submission (via console/edit_submission.html)
  console.views.copyedit_submission (via console/copyedit_submission.html)
  console.views.awaiting_authors (via console/awaiting_authors.html)
  console.views.publish_submission (via console/publish_submission.html)

Therefore, each of these views must determine the bulk URL prefix and
pass it as a template parameter.

('url_prefix' is not currently used here, but it *ought* to be used in
constructing the "anonymous access" URL; see commit
7ac6b5fd3691fda951f1eb6a06667c0792627b82.)

Commit:
Add AccessPolicy enum.

Commit:
Fix enum stuff.

Commit:
Remove PLATFORM_NAME from lightwave templates.

Commit:
Remove static content from share page.

Commit:
Change PLATFOR_NAME to SITE_NAME.

Commit:
Fix code style.

tompollard committed:
Merge pull request #1379 from upsidelab/feature/object_storage

GCS Object Storage

Commit:
Remove arguments.

tompollard committed:
Merge pull request #1438 from upsidelab/fix/remove-arguments

Remove arguments.

Commit:
Fix broken stuff.

Commit:
Add a new test case.

Commit:
Format the code.

Commit:
Remove more hard-coded values.

Commit:
Format the code.

Commit:
Remove circular import.

Commit:
Ignore F403 rule.

grzegorz-wcislo committed:
Only wait for GCS_HOST if set

grzegorz-wcislo committed:
Change unix socket to uwsgi tcp socket in docker

Commit:
Change Enum to IntEnum, use enum values instead of hard-coded ones.

tompollard committed:
Merge pull request #1437 from upsidelab/feature/access-policy-enum

Add access policy enum

tompollard committed:
Merge pull request #1440 from upsidelab/fix/dev_entrypoint_gcs

dev-entrypoint.sh: only wait for GCS_HOST if set

tompollard committed:
Merge pull request #1441 from upsidelab/feature/uwsgi_docker_tcp_socket

Change unix socket to uwsgi tcp socket in docker

tompollard committed:
Merge pull request #1439 from upsidelab/feature/flake8-config

Ignore F403 rule.

Commit:
Improve Section methods.

Commit:
Move Section model to the physionet app.

Commit:
Remove AWS info from FILE-STORAGE.md

Commit:
Remove duplicated imports.

Commit:
Remove dependencies.

Commit:
Add environment variables.

Commit:
Format the code.

Commit:
Update dependencies.

Commit:
New GitHub workflow script: physionet-upgrade-test.yml.

This workflow script runs the test-upgrade.sh script to test live
upgradability from the "origin/production" branch to the current head.

(This duplicates some, but not all, of the functionality of
physionet-build-test.yml.  Ideally these two scripts should be
combined into one.)

briangow committed:
Updates challenge information on PhysioNet as requested by Emory team. Breaks information down into four pages. Links to the Emory Github Pages website

briangow committed:
Minor content update to /about/challenge page

Commit:
physionet.test_gcs: run tests only if GCS_HOST is defined.

In order to run these tests, a (dummy) GCS server must be running.
Since this isn't done automatically, these tests should not be run by
default.  Moreover, we shouldn't assume that the server is available
at the hostname "gcs" - that hostname doesn't exist unless you've
taken steps to create it.

Therefore, skip these tests unless the configuration variable GCS_HOST
is set explicitly (which is done in docker/docker-compose.test.yml),
as well as skipping the tests if TEST_GCS_INTEGRATION is false.

Commit:
Add missing space.

tompollard committed:
Merge pull request #1429 from upsidelab/feature/debranding

Debranding

briangow committed:
Minor content updates to challenge pages

briangow committed:
Update general challenge email address

tompollard committed:
Fix internal links used by the second level navigation menu

The page should include links that allow a user to click an item in the left menu and get taken to a section within the page. These links were missed on the publish page template.

briangow committed:
Merge pull request #1446 from MIT-LCP/fix_internal_link1

Commit:
Move footer data to environment variables.

tompollard committed:
Merge pull request #1444 from MIT-LCP/auto-test-upgrade

Run test-upgrade.sh as part of test suite

dependabot[bot] committed:
Bump django from 2.2.24 to 2.2.25

Bumps [django](https://github.com/django/django) from 2.2.24 to 2.2.25.
- [Release notes](https://github.com/django/django/releases)
- [Commits](https://github.com/django/django/compare/2.2.24...2.2.25)

---
updated-dependencies:
- dependency-name: django
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

briangow committed:
Remove unnecessary import. Update code style in a few places

briangow committed:
Merge pull request #1455 from MIT-LCP/newer_newer_challenges

Commit:
Add missing context variable, create EMAIL_SIGNAURE variable.

Commit:
Format the code.

tompollard committed:
Merge pull request #1456 from upsidelab/fix/add-missing-email-data

Add missing email data

Commit:
Fix overlapping logo.

Commit:
Combine context processors.

tompollard committed:
Merge pull request #1448 from upsidelab/feature/debranding

Debranding - footer data and overlapping logo

tompollard committed:
allow strapline to be configured.

The strapline on the homepage (The Research Resource for Complex Physiologic Signals) is specific to PhysioNet. Move this STRAPLINE setting to the environment file, so it can be configured alongside SITENAME.

tompollard committed:
add example values to default environment file.

tompollard committed:
make accessibility page optional, and add default.

tompollard committed:
add header to the config file.

briangow committed:
Merge pull request #1452 from MIT-LCP/strapline

tompollard committed:
Merge pull request #1454 from MIT-LCP/dependabot/pip/django-2.2.25

Bump django from 2.2.24 to 2.2.25

dependabot[bot] committed:
Bump sqlparse from 0.4.1 to 0.4.2

Bumps [sqlparse](https://github.com/andialbrecht/sqlparse) from 0.4.1 to 0.4.2.
- [Release notes](https://github.com/andialbrecht/sqlparse/releases)
- [Changelog](https://github.com/andialbrecht/sqlparse/blob/master/CHANGELOG)
- [Commits](https://github.com/andialbrecht/sqlparse/compare/0.4.1...0.4.2)

---
updated-dependencies:
- dependency-name: sqlparse
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

tompollard committed:
Merge pull request #1453 from MIT-LCP/dependabot/pip/sqlparse-0.4.2

Bump sqlparse from 0.4.1 to 0.4.2

fiedosiukr committed:
Add missing variables.

fiedosiukr committed:
Add allow_file_downloads to demo entities.

tompollard committed:
Merge pull request #1432 from MIT-LCP/bulk-hostname

Configurable hostname for bulk downloads

tompollard committed:
Merge pull request #1458 from upsidelab/feature/debranding

Emails - add missing variable

tompollard committed:
bump sqlparse version to 0.4.2. Ref #1453

the version update was missed by dependabot.

tompollard committed:
bump django version to 2.2.25. Ref #1454

the version update was missed by dependabot.

Commit:
Merge pull request #1459 into dev.

Dependabot recently opened pull requests to update Python versions of
Django (#1454) and sqlparse (#1453). For reasons that are unclear to
me, Dependabot didn't update requirements.txt to match.

This pull request bumps the versions of the two packages in
requirements.txt to match those in poetry.lock.

fiedosiukr committed:
Modify migrations.

tompollard committed:
Merge pull request #1412 from upsidelab/feature/gcs-direct-files-downloads

Feature/gcs direct files downloads

dependabot[bot] committed:
Bump django from 2.2.25 to 2.2.26

Bumps [django](https://github.com/django/django) from 2.2.25 to 2.2.26.
- [Release notes](https://github.com/django/django/releases)
- [Commits](https://github.com/django/django/compare/2.2.25...2.2.26)

---
updated-dependencies:
- dependency-name: django
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

tompollard committed:
bump django to 2.2.26

tompollard committed:
Merge pull request #1461 from MIT-LCP/dependabot/pip/django-2.2.26

Bump django from 2.2.25 to 2.2.26

Commit:
Add missing variable.

Commit:
Add missing template variable.

Commit:
Apply code review suggestions.

Commit:
Add dropzone library.
